### ISSUE TITLE
setstate-in-render warning with useToggleState

### ISSUE THREAD

**Issue by valotas** [adobe/react-spectrum#1440]
# 🐛 Bug Report 

Warning when `onChange` is provided to `useToggleButton`:

```
Warning: Cannot update a component (`App`) while rendering a different component (`ToggleButton`). To locate the bad setState() call inside `ToggleButton`, follow the stack trace as described in https://reactjs.org/link/setstate-in-render ToggleButton@https://ry8pr.csb.app/src/ToggleButton.js:26:31
```

## 🤔 Expected Behavior

No warning

## 😯 Current Behavior

It looks like the provided `onChange` is not called within a `useEffect` and if it updates the state of a parent component the warning pops up as described by [the react team](https://reactjs.org/blog/2020/02/26/react-v16.13.0.html#warnings-for-some-updates-during-render)

## 💁 Possible Solution

wrap the onChange call in a `useEffect`. Possible workaround would be to wrap the provided onChange in a setTimeout (although I do not know how exactly react handles updates).

## 🔦 Context

We have a toggle button that reacts onChange and this reaction causes more components to be changed.

We tried both of the solutions but at least for the `useEffect` based one, we need to introduce another state in our Toggle component just to trigger the useEffect on change which looks like an overkill. Yet it looks like the correct one since it lets react notify the parent component via onChange after the child one has been updated

## 💻 Code Sample

https://codesandbox.io/s/bold-smoke-ry8pr

**Comment by AndyOGo** [comment]
I would like to add that using `useState` hook in a component itself and in a parent component works with React's `onChange` synthetic event for `<input />`.

Example here: https://codesandbox.io/s/compassionate-rhodes-xut0p

I think the issue is related to the custom event system of `react-aria`.
At least for `setState` of class components react makes sure that it's update is async.
> Currently, `setState` is asynchronous inside event handlers.
https://reactjs.org/docs/faq-state.html#when-is-setstate-asynchronous

I couldn't find a clear statement about `useState` hook and events, but at least from my example above it seems React handles it there too.

**Comment by snowystinger** [comment]
Ah, this is caused by an issue we're trying to address in useControlledState I believe.
This change fixes it, but we're currently evaluating if this is the right change https://github.com/adobe/react-spectrum/pull/1375
@devongovett

**Comment by valotas** [comment]
Our current hack is the following hook:

```
export function useCallbackInEffectHack<T>(
  value: T,
  handler: Handler<T>
): [T, Handler<T>] {
  const [lastValue, setValue] = useState(value);

  // keep the given value in sync with the state one
  useEffect(() => setValue(value), [value]);

  // call given callback if the value not the same with the on in state
  useEffect(() => {
    if (lastValue !== value) {
      handler(lastValue);
    }
  }, [lastValue, handler, value]);

  const wrappedHanlder = useCallback((val: T) => setValue(val), [setValue]);

  return [lastValue, wrappedHanlder];
}

```

**Comment by AndyOGo** [comment]
@devongovett 
@snowystinger 
Is there any update?

**Comment by snowystinger** [comment]
We're unfortunately busy this week with a pretty big pointer bug, soonest we'll be able to look at this is probably next week.

**Comment by snowystinger** [comment]
Closing as this is now addressed by https://github.com/adobe/react-spectrum/issues/2320 where we'll be removing our usage of useControlledState callbacks

**Comment by snowystinger** [comment]
We've finally merged this, #2304 It'll be available starting in the nightlies. Let us know if you find anything wrong with it.
